client-mr/mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/CelebornShuffleConsumer.java

```diff
@@ -0,0 +1,173 @@
+    this.merger = createMergeManager(context);
+  }
+
+  // Merge mapOutput and spill in local disks if necessary
+  protected MergeManager<K, V> createMergeManager(ShuffleConsumerPlugin.Context context) {
+    return new MergeManagerImpl<K, V>(
+        reduceId,
+        mrJobConf,
+        context.getLocalFS(),
+        context.getLocalDirAllocator(),
+        reporter,
+        context.getCodec(),
+        context.getCombinerClass(),
+        context.getCombineCollector(),
+        context.getSpilledRecordsCounter(),
+        context.getReduceCombineInputCounter(),
+        context.getMergedMapOutputsCounter(),
+        this,
+        context.getMergePhase(),
+        context.getMapOutputFile());
+  }
+
+  private ShuffleClientMetrics createMetrics(
+      org.apache.hadoop.mapreduce.TaskAttemptID taskAttemptID, JobConf jobConf)
+      throws NoSuchMethodException, InvocationTargetException, InstantiationException,
+          IllegalAccessException {
+    // for hadoop 3
+    Method createMethod = null;
+    try {
+      ShuffleClientMetrics.class.getDeclaredMethod(
+          "create", org.apache.hadoop.mapreduce.TaskAttemptID.class, JobConf.class);
+    } catch (Exception e) {
+      // ignore this exception because
+    }
+    if (createMethod != null) {
+      return (ShuffleClientMetrics) createMethod.invoke(null, taskAttemptID, jobConf);
+    }
+    // for hadoop 2
+    Constructor constructor =
+        ShuffleClientMetrics.class.getDeclaredConstructor(
+            org.apache.hadoop.mapreduce.TaskAttemptID.class, JobConf.class);
+    constructor.setAccessible(true);
+    return (ShuffleClientMetrics) constructor.newInstance(taskAttemptID, jobConf);
+  }
+
+  @Override
+  public RawKeyValueIterator run() throws IOException, InterruptedException {
+    logger.info(
+        "In reduce:{}, Celeborn mr client start to read shuffle data."
+            + " Create inputstream with params: shuffleId 0 reduceId:{} attemptId:{}",
+        reduceId,
+        reduceId.getTaskID().getId(),
+        reduceId.getId());
+
+    CelebornInputStream shuffleInputStream =
+        shuffleClient.readPartition(
+            0, reduceId.getTaskID().getId(), reduceId.getId(), 0, Integer.MAX_VALUE);
+    CelebornShuffleFetcher<K, V> shuffleReader =
+        new CelebornShuffleFetcher(
+            reduceId, taskStatus, merger, copyPhase, reporter, metrics, shuffleInputStream);
+    shuffleReader.fetchAndMerge();
+
+    copyPhase.complete();
+    taskStatus.setPhase(TaskStatus.Phase.SORT);
+    reduceTask.statusUpdate(umbilical);
+
+    // Finish the on-going merges...
+    RawKeyValueIterator kvIter = null;
+    try {
+      kvIter = merger.close();
+    } catch (Throwable e) {
+      throw new ShuffleError("Error while doing final merge ", e);
+    }
+
+    logger.info("In reduce: " + reduceId + ", Celeborn mr client read shuffle data complete");
+
+    return kvIter;
+  }
+
+  @Override
+  public void close() {}
```

Comment:
Does the `shuffleClient` need to close?

Reply:
`shuffleClient` is a singleton. It can not close.